packages/react-dom/src/fire/ReactFireEvents.js

```diff
@@ -0,0 +1,322 @@
+  propName: string,
+) {
+  const isDocumentOrFragment =
+    rootContainerElement.nodeType === DOCUMENT_NODE ||
+    rootContainerElement.nodeType === DOCUMENT_FRAGMENT_NODE;
+  const doc = isDocumentOrFragment
+    ? rootContainerElement
+    : rootContainerElement.ownerDocument;
+  const isPolyfilledEvent = polyfilledEvents.hasOwnProperty(propName);
+
+  if (isPolyfilledEvent) {
+    listenToPolyfilledEvent(doc, propName);
+  } else {
+    const eventName = normalizeEventName(propName);
+    listenTo(eventName, doc);
+  }
+}
+
+function proxyInteractiveListener(eventName, containerDomNode, event) {
+  interactiveUpdates(proxyListener, eventName, containerDomNode, event);
+}
+
+function createProxyContext(
+  containerDomNode,
+  event,
+  eventName,
+  eventTarget,
+  ancestors,
+) {
+  return {
+    ancestors,
+    containerDomNode,
+    currentTarget: null,
+    defaultPrevented: false,
+    event,
+    eventName,
+    eventTarget,
+    fiber: null,
+    isClickEvent: eventName === CLICK || name === DOUBLE_CLICK,
+  };
+}
+
+function dispatchEvent(proxyContext) {
+  const {ancestors, containerDomNode, eventName, eventTarget} = proxyContext;
+  if (ancestors.length === 0) {
+    if (eventName === 'mouseout') {
+      dispatchPolyfills(containerDomNode, eventTarget, proxyContext);
+    }
+    return;
+  }
+  for (let x = 0; x < ancestors.length; x++) {
+    const ancestor = ancestors[x];
+    const targetDomNode = getDOMNodeFromFiber(ancestor);
+    proxyContext.fiber = ancestor;
+    // Dispatch run all non-polyfilled events first using two phase traversal
+    traverseTwoPhase(proxyContext);
+    // Then dispatch all polyfilled events (onChange, onBeforeInput etc).
+    // Each polyfilled event has as its own event handler that provides the
+    // dispatch mechanism to use.
+    dispatchPolyfills(containerDomNode, targetDomNode, proxyContext);
+  }
+}
+
+export function proxyListener(eventName, containerDomNode, event) {
+  if (!eventsEnabled) {
+    return;
+  }
+  const eventTarget = getEventTarget(event);
+  let targetFiber = getClosestFiberFromDOMNode(eventTarget);
+
+  if (
+    targetFiber !== null &&
+    typeof targetFiber.tag === 'number' &&
+    !isFiberMounted(targetFiber)
+  ) {
+    // If we get an event (ex: img onload) before committing that
+    // component's mount, ignore it for now (that is, treat it as if it was an
+    // event on a non-React tree). We might also consider queueing events and
+    // dispatching them after the mount.
+    targetFiber = null;
+  }
+  const ancestors = getEventTargetAncestorFibers(targetFiber);
+  const proxyContext = createProxyContext(
+    containerDomNode,
+    event,
+    eventName,
+    eventTarget,
+    ancestors,
+  );
+
+  monkeyPatchNativeEvent(event, proxyContext);
+  batchedUpdates(dispatchEvent, proxyContext);
+}
+
+function stopPropagation() {
+  this.cancelBubble = true;
+  if (!this.immediatePropagationStopped) {
+    this.stopImmediatePropagation();
```

Comment:
why does stopPropagation call stopImmediatePropagation (also IE support?)

Reply:
Was also taken from Inferno, need to check if we still need to do this, but last time I checked it was to get around some issues in IE.